### PR TITLE
feat: sat collision algorithm for convex polygons and circles

### PR DESCRIPTION
**Pull request by WeRiano**
Added collision detection algorithm for 2D shapes (convex polygons and circles) that uses [SAT](https://en.wikipedia.org/wiki/Hyperplane_separation_theorem)

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes:

### CODE REVIEW

**Comment by WeRiano** [comment]
linked to [#1955](https://github.com/TheAlgorithms/C-Plus-Plus/issues/1955)

File: geometry/sat_collision.cpp [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+/*********************************** MAIN *************************************/
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main() {
+    test();
+    return 0;
+}
+/******************************************************************************/
```
Comment (Panquesito7) [inline geometry/sat_collision.cpp:200]: Missing an endline. Without this endline, there could be issues with some IDEs and configurations.

```suggestion
/******************************************************************************/

```


__
[sema-logo](https://semasoftware.com/gh) &nbsp;**Summary:** :hammer_and_wrench: This code needs a fix&nbsp; | &nbsp;**Tags:** Inefficient

File: geometry/sat_collision.h [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+* The algorithm should run in O(n) time for polygon-X collisions and O(1) for
+* circle-circle collisions, where n is the largest number of vertices that
+* belongs to one of the polygons
+
+* @author [William Fors](https://github.com/WeRiano)
+*/
+
+namespace geometry {
+
+namespace collision {
```
Comment (Panquesito7) [inline geometry/sat_collision.h:26]:
```suggestion
/**
 * @namespace geometry
 * @brief Geometry-related algorithms
 */
namespace geometry {
/**
 * @namespace collision
 * @brief Functions for the [Hyperplane Separation Theorem]
* (https://en.wikipedia.org/wiki/Hyperplane_separation_theorem) algorithm implementation
*/
namespace collision {
```


__
[sema-logo](https://semasoftware.com/gh) &nbsp;**Summary:** :hammer_and_wrench: This code needs a fix

File: geometry/sat_collision.h [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+
+/**
+ * @brief Performs the SAT collision algorithm on two 2-dimensional shapes
+ * @param shape1 First shape
+ * @param shape2 Second shape
+ * @return `true` if the shapes collide.
+ * @return `false` otherwise
+ */
+bool sat(const Shape& shape1, const Shape& shape2);
+}}
```
Comment (Panquesito7) [inline geometry/sat_collision.h:251]:
```suggestion
}  // namespace collision
}  // namespace geometry

```


__
[sema-logo](https://semasoftware.com/gh) &nbsp;**Summary:** :hammer_and_wrench: This code needs a fix&nbsp; | &nbsp;**Tags:** Inelegant

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Amazing work! 🚀
P.S.: Sorry for the delayed response.

**Pull request closed without merging**